core/src/main/scala/kafka/raft/KafkaFuturePurgatory.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.raft
+
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.atomic.AtomicBoolean
+
+import kafka.server.{DelayedOperation, DelayedOperationPurgatory}
+import kafka.utils.Logging
+import kafka.utils.timer.Timer
+import org.apache.kafka.common.errors.TimeoutException
+import org.apache.kafka.raft.FuturePurgatory
+
+/**
+ * Simple purgatory shim for integration with the Raft library. We assume that
+ * both [[await()]] and [[completeAll()]] are called in the same thread.
+ */
+class KafkaFuturePurgatory(brokerId: Int, timer: Timer, reaperEnabled: Boolean = true)
+  extends FuturePurgatory[Unit] with Logging {
+
+  private val key = new Object()
+  private val purgatory = new DelayedOperationPurgatory[DelayedRaftRequest](
+    "raft-request-purgatory", timer, brokerId, reaperEnabled = reaperEnabled)
+
+  override def await(future: CompletableFuture[Unit], maxWaitTimeMs: Long): Unit = {
+    val op = new DelayedRaftRequest(future, maxWaitTimeMs)
+    purgatory.tryCompleteElseWatch(op, Seq(key))
+    op.isCompletable.set(true)
+  }
+
+  override def completeAll(value: Unit): Unit = {
+    purgatory.checkAndComplete(key)
+  }
+
+  override def numWaiting(): Int = {
+    purgatory.numDelayed
+  }
+}
+
+class DelayedRaftRequest(future: CompletableFuture[Unit], delayMs: Long)
+  extends DelayedOperation(delayMs) {
+
+  val isCompletable = new AtomicBoolean(false)
+  val isExpired = new AtomicBoolean(false)
+
+  override def onExpiration(): Unit = {}
```

Comment:
The semantics of `DelayedOperation` are a little odd. Regardless of expiration, the `onComplete` callback _always_ gets fired. The `onExpiration` callback may also get fired if the operation timed out, but only after `onComplete`. The `isExpired` flag here and the `run()` override below are my workaround to get the behavior I wanted with these semantics, but let me know if you can think of a better idea.

Reply:
I think I understood the motivation here. Although this still looks weird, I agree there is no better solution as of now. A slightly different approach would be trying to utilize the `onExpiration` and bypass the `onComplete` during timeout:
```
 override def onExpiration(): Unit = {
    future.completeExceptionally(new TimeoutException("Request timed out in purgatory"))
  }

  override def onComplete(): Unit = {
    if (!isExpired.get())
      future.complete(())
  }
```
I'm not suggesting any real change here, but this feel more natural to me subjectively.

Reply:
Hmm.. I was going to make this change, but I think it might not be safe. The logic on expiration is basically this:
```
1.    isExpired.set(true)
2.    if (forceComplete())
3.      onExpiration()
```
If another thread calls `forceComplete` between 1) and 2), then the future would never get completed.